### PR TITLE
Data coordination and selective visualization of edges

### PR DESCRIPTION
**Pull request by woodaaron**
Introduces a data coordinator class to manage surface features within the surface detection node.
Updates the primary point type used for segmentation and planning to include color data.
Includes updates to UI which allow a user to selectively visualize edge poses (can visualize one edge at a time).

### CODE REVIEW

File: godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+      id_counter_ = 0;
+      if(records_.size() > 0)
+        records_.clear();
+    }
+
+    /**
+     * @brief init Initializes DataCoordinator object
+     * @return success of operation
+     */
+    bool init()
```
Comment (Jmeyer1292) [inline godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h]: Why both a meaningful constructor and the `init()` function? Perhaps you have a reset function that I haven't gotten to yet, but if that's the case then why not have `DataCoordinator()` dispatch to `init()`?

File: godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+      ss << "]";
+      return ss.str();
+    }
+
+
+  public:
+    //! Default Constructor
+    DataCoordinator()
+    {
+      id_counter_ = 0;
```
Comment (Jmeyer1292) [inline godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h]: You set this to zero here, but you also listed a default initializer above.

File: godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+
+    /**
+     * @brief getCloud Returns a cloud of interest
+     * @param type Type of cloud to return (currently implemented: input, surface)
+     * @param id ID of the desired record
+     * @param cloud Destination for cloud
+     * @return true if record is found and type is valid, false otherwise
+     */
+    bool getCloud(int type, int id, pcl::PointCloud<pcl::PointXYZRGB>& cloud)
```
Comment (Jmeyer1292) [inline godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h]: Perhaps type should be an `enum class` for safety purposes.

Comment (Jmeyer1292) [inline godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h]: Can we make this function `const`?

File: godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+      return false;
+    }
+
+
+    /**
+     * @brief getSurfaceName Returns the surface name from a record
+     * @param id ID of the desired record
+     * @param name Destination variable for surface name
+     */
+    void getSurfaceName(int id, std::string& name)
```
Comment (Jmeyer1292) [inline godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h]: You have a `bool someAction()` style interface for other functions. Could this function fail to find said `id` and can we communicate that to the user?

File: godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    {
+      for(auto& rec : records_)
+      {
+        if(id == rec.id_)
+        {
+          mesh = rec.surface_mesh_;
+          break;
+        }
+      }
+    }
```
Comment (Jmeyer1292) [inline godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h]: Again, what if this fails?

File: godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+
+    /**
+     * @brief getEdgePosesByName Obtains PoseArray for edge with name edge_name
+     * @param edge_name
+     * @param edge_poses Destination for PoseArray
+     * @return true if record is found and an edge with edge_name exists within the record, false otherwise
+     */
+    bool getEdgePosesByName(const std::string& edge_name, geometry_msgs::PoseArray& edge_poses)
```
Comment (Jmeyer1292) [inline godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h]: This function makes me think a little about the implicit assumptions of this class:
 1. Does every edge name have to be globally unique (i.e. part 1 and part can not both have "edge1")?
 2. If this is the case, should we enforce this when we insert a new edge?
 3. This function here appears to grab the first record with any edge of the given name: is this expected behaviour?

**Review by Jmeyer1292 — COMMENTED** [review]
More comments to come

File: godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h (hunk 1 above)
Comment (woodaaron) [reply godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h]: It was intended more as a way to clean the object (not sure I had a good purpose there) and to mimic the style of InteractiveSurfaceServer, SurfaceBlendingService, and RobotScan. 

The other 3 classes call init() in surface_blending_service.cpp

`if (surface_detection_.init() && robot_scan_.init() && surface_server_.init() && data_coordinator_.init())
  {...`

And having the init() signals to the reader that the object is ready for use. It is, however, logically superfluous, as I don't use it elsewhere, and I can remove it with few changes.

File: godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h (hunk 6 above)
Comment (woodaaron) [reply godel_surface_detection/include/godel_surface_detection/data_coordinator/data_coordinator.h]: Fair point. As written, there is an implicit requirement that edges have unique names (effectively local to the surface right now, but potentially global later). This allows the UI to request visualization of an edge in a single service call with looser coupling to the surface_detection node.

Edge names are implemented as surface names suffixed with "_edge" and an increment (e.g. 1_edge3). We don't currently allow changes to the edge name through the UI, and renameEdges was intended to work in concert with an attempt to rename a surface. However, we could get right back into a situation where our default edge identifier is ambiguous if we expand what we allow the user to do.

Best plan might be to add an id field for edges then pass the field to the UI and let it coordinate the string it presents to the user with the unique id of the edge.

**Comment by Jmeyer1292** [comment]
+1
@woodaaron Please open issues per our discussion.

**Pull request merged**